File: app/mlboard/api/services/maintenance.py
from mlboard.core import models as ms
from mlboard.core import queries as qs
from mlboard.core import db
from fastapi import APIRouter, Query
import typing
from logging import getLogger
from cytoolz.curried import pipe, map
from pydantic import BaseModel
import uuid
from datetime import datetime

router = APIRouter()


class MaintenanceForm(BaseModel):
    id: uuid.UUID
    name: str
    chamber_id: uuid.UUID


@router.get('/maintenance/all')
async def all():
    async with db.get_conn() as conn:
        rows = await qs.Maintenance(conn).all()
    return rows

@router.get('/maintenance')
async def get(id: uuid.UUID):
    async with db.get_conn() as conn:
        row = await qs.Maintenance(conn).get_by(id=id)
    return row


@router.get('/maintenance/filter-in')
async def filter_by(
    chamber_ids: typing.List[uuid.UUID] = Query([]),
):
    async with db.get_conn() as conn:
        rows = await qs.Maintenance(conn).filter_in(chamber_ids)
    return rows


@router.delete('/maintenance', response_model=uuid.UUID)
async def delete_by(id: uuid.UUID):
    async with db.get_conn() as conn:
        queried_id = await qs.Maintenance(conn).delete_by(id=id)
    return queried_id


@router.post('/maintenance', response_model=uuid.UUID)
async def upsert(payload: MaintenanceForm):
    maintenance = ms.Maintenance(**dict(payload))
    async with db.get_conn() as conn:
        queried_id = await qs.Maintenance(conn).upsert(maintenance)
    return queried_id

File: board/src/core/api/maintenance.ts
import { IMaintenance } from "~/core/models"
import {  StatusLevel } from '~/core/enums'
import axios from 'axios'

interface IMaintenanceResponse {
  id: string
  name: string
  chamber_id: string
  value: number
  collect_date: string
  status:  StatusLevel
}

interface IMaintenanceForm {
  id: string
  name: string
  chamberId: string
}

export const filterIn = async (chamberIds:string[]):Promise<IMaintenance[]> => {
  const res = await axios.get('api/maintenance/filter-in', {
    params: {
      chamber_ids: chamberIds
    }
  })
  const resData = res.data as IMaintenanceResponse[]
  return resData.map(x => { 
    return {
      id: x.id,
      name: x.name,
      chamberId: x.chamber_id,
      value:  x.value,
      collectDate: x.collect_date,
      status: x.status,
    }
  })
}

export const get = async (id:string):Promise<IMaintenance|undefined> => {
  const res = await axios.get('api/maintenance', {
    params: {
      id: id
    }
  })
  const resData = res.data as IMaintenanceResponse | undefined
  if(resData){
    return {
      id: resData.id,
      name: resData.name,
      chamberId: resData.chamber_id,
      value:  resData.value,
      collectDate: resData.collect_date,
      status: resData.status,
    }
  }
}

export const upsert =  async (payload: IMaintenanceForm):Promise<string> => {
  const res = await axios.post('api/maintenance', {
    id: payload.id,
    name: payload.name,
    chamber_id: payload.chamberId,
  })
  const resData = res.data as string
  return resData
}

export const deleteBy = async (id:string):Promise<string> => {
  const res = await axios.delete('api/maintenance', {
    params: {
      id: id
    }
  })
  const resData = res.data as string
  return resData
}
